### PR TITLE
Extend costmap_2d interface with methods required by Hybrid A* planner

### PR DESCRIPTION
**Pull request by corot**
Required by https://dev.azure.com/rapyuta-robotics/flappter/_workitems/edit/246

https://github.com/open-navigation/navigation2/tree/main/nav2_smac_planner uses some methods from nav2's costmap_2d missing on ROS 1. As the implementation of all of them is trivial, and potentialy useful in other cases, I preferred to extend our costmap_2d (already different from upstream one) to adapt and make more complicated nav2_smac_planner's code.

### CODE REVIEW

File: costmap_2d/src/costmap_2d.cpp [hunk 1]
```diff
@@ -219,6 +224,23 @@ bool Costmap2D::worldToMap(double wx, double wy, unsigned int& mx, unsigned int&
...
 
+bool Costmap2D::worldToMapContinuous(double wx, double wy, float& mx, float& my) const
+{
+  if (wx < origin_x_ || wy < origin_y_)
+  {
+    return false;
+  }
+
+  mx = static_cast<float>((wx - origin_x_) / resolution_) + 0.5f;
+  my = static_cast<float>((wy - origin_y_) / resolution_) + 0.5f;
```
Comment (akela1101) [inline costmap_2d/src/costmap_2d.cpp:235]: Why `+ 0.5f` ?

File: costmap_2d/include/costmap_2d/footprint.h [hunk 2]
```diff
@@ -121,7 +121,7 @@ bool makeFootprintFromString(const std::string& footprint_string, std::vector<ge
  * @brief Read the ros-params "footprint" and/or "robot_radius" from
  * the given NodeHandle using searchParam() to go up the tree.
  */
-std::vector<geometry_msgs::Point> makeFootprintFromParams(ros::NodeHandle& nh);
+std::pair<std::vector<geometry_msgs::Point>, bool> makeFootprintFromParams(ros::NodeHandle& nh);
```
Comment (akela1101) [inline costmap_2d/include/costmap_2d/footprint.h]: please, document the return

**Review by akela1101 — APPROVED** [review]

File: costmap_2d/src/costmap_2d.cpp (hunk 1 above)
Comment (corot) [reply costmap_2d/src/costmap_2d.cpp:235]: To use the center of the cell, instead of the border. 
disclaimer: It's not my choice,,, I copied from nav2

Comment (kota-rr) [reply costmap_2d/src/costmap_2d.cpp:235]: please write your reply above to the code as a comment. 
people will start :thinking: in 2026 again.

File: costmap_2d/include/costmap_2d/costmap_2d_ros.h [hunk 3]
```diff
@@ -235,6 +235,14 @@ class Costmap2DROS
    * getUnpaddedRobotFootprint(). */
   void setUnpaddedRobotFootprintPolygon(const geometry_msgs::Polygon& footprint);
 
+  /**
+   * @brief  Get the costmap's use_radius_ parameter, corresponding to
+   * whether the footprint for the robot is a circle with radius robot_radius_
+   * or an arbitrarily defined footprint in footprint_.
+   * @return  use_radius_
+   */
+  bool getUseRadius() { return use_radius_; }
```
Comment (kota-rr) [inline costmap_2d/include/costmap_2d/costmap_2d_ros.h]: ` bool getUseRadius() const`

File: costmap_2d/src/footprint.cpp [hunk 4]
```diff
@@ -246,7 +246,7 @@ std::vector<geometry_msgs::Point> makeFootprintFromParams(ros::NodeHandle& nh)
   // Else neither param was found anywhere this knows about, so
   // defaults will come from dynamic_reconfigure stuff, set in
   // cfg/Costmap2D.cfg and read in this file in reconfigureCB().
-  return points;
+  return { points, true };
```
Comment (kota-rr) [inline costmap_2d/src/footprint.cpp]: Q: isn't `use_radius` `false` if "robot_radius" doesn't exist?

**Review by kota-rr — APPROVED** [review]

File: costmap_2d/src/footprint.cpp (hunk 4 above)
Comment (corot) [reply costmap_2d/src/footprint.cpp]: right

File: costmap_2d/include/costmap_2d/footprint.h [hunk 5]
```diff
@@ -120,8 +120,10 @@ bool makeFootprintFromString(const std::string& footprint_string, std::vector<ge
 /**
  * @brief Read the ros-params "footprint" and/or "robot_radius" from
  * the given NodeHandle using searchParam() to go up the tree.
+ * @return The footprint as a list of points and whether we used the robot_radius.
+ * I that's the case, the footprint will be a circle of "robot_radius".
```
Comment (renan028) [inline costmap_2d/include/costmap_2d/footprint.h]:
```suggestion
 * If that's the case, the footprint will be a circle of "robot_radius".
```

File: costmap_2d/include/costmap_2d/footprint.h [hunk 6]
```diff
@@ -120,8 +120,10 @@ bool makeFootprintFromString(const std::string& footprint_string, std::vector<ge
...
  */
-std::vector<geometry_msgs::Point> makeFootprintFromParams(ros::NodeHandle& nh);
+std::pair<std::vector<geometry_msgs::Point>, bool> makeFootprintFromParams(ros::NodeHandle& nh);
```
Comment (renan028) [inline costmap_2d/include/costmap_2d/footprint.h]: Q: do we need to make this change? 
we can retrieve the extra info (pair.second) by calling `getUseRadius`, right?

Comment (corot) [reply costmap_2d/include/costmap_2d/footprint.h]: yes, but I thought it was more natural this way

**Review by tomin11 — APPROVED** [review]

File: costmap_2d/include/costmap_2d/footprint.h (hunk 6 above)
Comment (vsamy) [reply costmap_2d/include/costmap_2d/footprint.h]: `std::optional` maybe fits better here?

File: costmap_2d/src/costmap_2d.cpp [hunk 7]
```diff
@@ -219,6 +224,23 @@ bool Costmap2D::worldToMap(double wx, double wy, unsigned int& mx, unsigned int&
...
+
+  if (mx < size_x_ && my < size_y_)
+  {
+    return true;
+  }
+  return false;
```
Comment (vsamy) [inline costmap_2d/src/costmap_2d.cpp]:
```suggestion
  return mx < size_x_ && my < size_y_;
```
You litteraly wrote if true return true if false return false. XD

File: costmap_2d/src/footprint.cpp [hunk 8]
```diff
@@ -209,7 +209,7 @@ bool makeFootprintFromString(const std::string& footprint_string, std::vector<ge
 
 
 
-std::vector<geometry_msgs::Point> makeFootprintFromParams(ros::NodeHandle& nh)
+std::pair<std::vector<geometry_msgs::Point>, bool> makeFootprintFromParams(ros::NodeHandle& nh)
```
Comment (vsamy) [inline costmap_2d/src/footprint.cpp]: S: Prefer `std::optional` for this usecase

File: costmap_2d/include/costmap_2d/footprint.h (hunk 6 above)
Comment (corot) [reply costmap_2d/include/costmap_2d/footprint.h]: mmm...  I already changed on the client side to use this,,, will keep it as it is

**Review by renan028 — APPROVED** [review]

File: costmap_2d/src/footprint.cpp (hunk 8 above)
Comment (corot) [reply costmap_2d/src/footprint.cpp]: right, we make a .deb for this pkg, so returning a pair would have forced my to path all existing releases

changed, but using pointer

**Pull request merged**